pkg/controller/job/job_controller.go

```diff
@@ -1090,6 +1110,32 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 	return nil
 }
 
+// canRemoveFinalizer determines if the pod's finalizer can be safely removed.
+// The finalizer can be removed when:
+//   - the entire Job is terminating; or
+//   - the pod's index is succeeded; or
+//   - the Pod is considered failed, unless it's removal is delayed for the
+//     purpose of transferring the JobIndexFailureCount annotations to the
+//     replacement pod. the entire Job is terminating the finalizer can be
+//     removed unconditionally.
+func canRemoveFinalizer(logger klog.Logger, job *batch.Job, finishedCond *batch.JobCondition, pod *v1.Pod, considerPodFailed bool, podsWithDelayedDeletionPerIndex map[int]*v1.Pod) bool {
+	if job.DeletionTimestamp != nil || finishedCond != nil || pod.Status.Phase == v1.PodSucceeded {
+		return true
+	}
+	if considerPodFailed {
+		if hasBackoffLimitPerIndex(job) {
+			if index := getCompletionIndex(pod.Annotations); index != unknownCompletionIndex {
+				if p, ok := podsWithDelayedDeletionPerIndex[index]; ok && p.UID == pod.UID {
```

Comment:
> IIUC, we want to know if this pod has the highest failure count for the index and is not active.

This is correct, and in case of a tie, I discriminate also by `finishTime`. The tie may happen due to Ignore pod failure policy rule. Still, I could easily extend the sorting function for pods.

> for the first check, we could just sort the pods by failure count in addition to index, at the beginning of trackJobStatusAndRemoveFinalizers. For the second, we would have to keep the set of active indexes. Would that save some memory or simplify code?

One issue is that I need this information in two places: `trackJobStatusAndRemoveFinalizers` and `manageJob` to determine if the index can or cannot already by recreated. I wouldn't need it for manage job if we only had a global backoff limit delay. But currently, I use the information to get the finish time to compute the delay per index. 
If we go for just global backoff delay, then yes, the new sorting + activeIndexes should be enough.

Reply:
> One issue is that I need this information in two places: trackJobStatusAndRemoveFinalizers and manageJob 

Right, I noticed after I left this comment. Still worth rationalizing the points. Thanks for clarifying :)

> but it seems building such a map should still should be negligible compared to sending api requests.

At some point in kube-scheduler, it was slower to build some critical maps than to send API requests. So we replaced them with slices or other strategies where possible.

Hard to know if we are crossing a threshold here. It will highly depend on parallelism. We might need to re-evaluate for beta based on integration or e2e level benchmarks.

Reply:
Yeah, hard to tell just by static code inspection. Added a point for Beta graduation to test performance in the KEP update: https://github.com/kubernetes/enhancements/pull/4123